### ISSUE TITLE
Allow @GeneratedValue for non-@Id fields

### ISSUE THREAD

**Issue by lukasj** [eclipse-ee4j/jpa-api#113]
Hello,

My scenario is the following - I have a <tt>@MappedSuperclass</tt> which defines a field <tt>pk</tt> (which is generated value with a custom generic generator) but I also have an entity called <tt>Order</tt> with a property <tt>code</tt>. I have created a custom generic generator which calculates on the server what the order code must be, but even though I have specified it as this:

```
@Column(name = "code", nullable = false)
    @GenericGenerator(name = "uuid", strategy = "uuid2")
    @GeneratedValue(generator = "uuid")
    private String code;
```

This generator is never called. I checked the javadoc and there explicitly says that the <tt>@GeneratedValue</tt> is used together with <tt>@Id</tt>. My proposal is to allow it to be used on columns that are not <tt>@Id</tt>.

**Comment by thejeff77** [comment]
Is there any update on this? The annotations still don't seem to be working without additional annotations such as Id. I've tried options from the post below excluding extra Id or column annotations and the field is not generated. 

https://stackoverflow.com/questions/47934661/generate-unpredictable-random-id-for-entity

**Comment by sumitdhungana14** [comment]
Do we have any updates on this? `@GenericGenerator` doesn't generate values for columns other than the `@Id`ed column.

**Comment by sebersole** [comment]
I think this would be a good addition.  But if not already a separate request, I suggest we consider how this plays with update/insert handling.  I assume the intention here is specifically for insertion-time generation.  However, update-time generators are plenty useful as well - modified-timestamp e.g.

I'd actually suggest not re-using `@GeneratedValue` here, which is really a shame because that name really fits better here for this non-id case.  But regardless, for the sake of discussion let's call it `@GeneratedNonIdValue`... something like:

```
@interface GeneratedNonIdValue {
    enum Timing { INSERT, UPDATE, BOTH }

    Timing timing() default BOTH;
    GenerationType strategy() default AUTO;
    String generator() default "";
}
```

**Comment by asariev** [comment]
Agree with @sebersole!

**Comment by gavinking** [comment]
> I think this would be a good addition. But if not already a separate request, I suggest we consider how this plays with update/insert handling. I assume the intention here is specifically for insertion-time generation. However, update-time generators are plenty useful as well

Indeed. If we do this, we should go the whole hog and do this issue, #342, and update-time generation all at once.

Note that there's an existing issue, #51, which proposes using `@PrePersist` and `@PreUpdate` annotations to identify insert-time and update-time generation. Since we actually already have those annotations, that's actually a pretty interesting idea, and I sorta like it.

```
@PrePersist @GeneratedValue Instant created;
@PrePersist @PreUpdate @GeneratedValue Instant updated;
```

That's not awful at all. (It's superficially a bit different to the track we took with the generator framework in Hibernate, but it's not at all inconsistent with the ideas there.)

Anyway, I actually want to close #51 because the proposal there is written up in a very weird way, and it's had no activity, but I'm recording the idea here, since it's a possible approach we could take if/when we decide to address this issue, probably in conjunction with #342.

**Comment by hendisantika** [comment]
This is still open ya?

Or is it closed?

How do we fix it?

Thanks

**Comment by rikes** [comment]
In my case I needed to include a field whose value was generated by a sequence in the Oracle database.

Adding _@Generated_ from Hibernate worked

```
@Generated(GenerationTime.INSERT)
@Column(name = "my_field")
private Long field;
```

----
Source:
https://stackoverflow.com/a/23831524

**Comment by gavinking** [comment]
I don't see this happening in JPA 4 (apologies to the 22 people who voted for this very popular issue). But I think it would be easy for this one to make it to the top of the priority list for 4.1.
